components/validator/challenger.go

```diff
@@ -65,191 +93,413 @@ func NewChallenger(ctx context.Context, cfg Config, l log.Logger) (*Challenger,
-		log: l,
-		cfg: cfg,
-		ctx: ctx,
+		log:  l,
+		cfg:  cfg,
+		ctx:  ctx,
+		done: make(chan struct{}),
 
 		callOpts: utils.NewCallOptsWithSender(ctx, cfg.TxManager.From()),
 		txOpts:   utils.NewSimpleTxOpts(ctx, cfg.TxManager.From(), cfg.TxManager.Signer),
 
+		l1Client: cfg.L1Client,
+
 		l2ooContract:      l2ooContract,
+		l2ooABI:           l2ooABI,
 		colosseumContract: colosseumContract,
+		colosseumABI:      colosseumABI,
 
 		submissionInterval:        submissionInterval,
 		finalizationPeriodSeconds: finalizationPeriodSeconds,
 		l2BlockTime:               l2BlockTime,
-	}, nil
-}
 
-func (c *Challenger) IsChallengeInProgress() (bool, error) {
-	return c.colosseumContract.IsInProgress(c.callOpts)
+		txCandidatesChan: txCandidateChan,
+	}, nil
 }
 
-func (c *Challenger) GetChallengeInProgress() (bindings.TypesChallenge, error) {
-	return c.colosseumContract.GetChallengeInProgress(c.callOpts)
-}
+func (c *Challenger) Start() error {
+	c.log.Info("start challenger")
 
-func (c *Challenger) OutputAtBlockSafe(blockNumber uint64, includeNextBlock bool) (*eth.OutputResponse, error) {
-	ctx, cancel := context.WithTimeout(c.ctx, c.cfg.NetworkTimeout)
-	defer cancel()
-	output, err := c.cfg.RollupClient.OutputAtBlock(ctx, blockNumber, includeNextBlock)
+	// set the challenger's checkpoint to the latest output index
+	// because when challenger starts, it is guaranteed that the past outputs are checked through scanPrevOutputs
+	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(c.callOpts)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	c.checkpoint = new(big.Int).Sub(nextOutputIndex, common.Big1)
 
-	return output, nil
+	// scan all the outputs within the finalization window
+	c.wg.Add(1)
+	go c.scanPrevOutputs()
+
+	// if challenge mode on, subscribe L2 output submission events
+	if !c.cfg.ChallengerDisabled {
+		c.wg.Add(1)
+		go c.subscribeL2OutputSubmitted()
+	}
+
+	// subscribe challenge creation events
+	c.wg.Add(1)
+	go c.subscribeChallengeCreated()
+
+	return nil
 }
 
-type OutputRange struct {
-	OutputIndex *big.Int
-	StartBlock  uint64
-	EndBlock    uint64
+func (c *Challenger) Stop() {
+	c.log.Info("stop challenger")
+	close(c.done)
+	c.wg.Wait()
 }
 
-func (c *Challenger) GetInvalidOutputRange() (*OutputRange, error) {
-	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(c.callOpts)
-	if err != nil {
-		return nil, err
-	}
-	latestOutputIndex := new(big.Int).Sub(nextOutputIndex, common.Big1)
-	if latestOutputIndex.Cmp(common.Big1) == -1 {
-		c.log.Info("challenge for genesis output is not allowed", "latestOutputIndex", latestOutputIndex)
-		return nil, nil
+// scanPrevOutputs scans all the previous outputs within the finalization window.
+// If there are invalid outputs, create challenge.
+// If there are challenges in progress, keep handling them.
+func (c *Challenger) scanPrevOutputs() {
+	defer c.wg.Done()
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), c.cfg.NetworkTimeout)
+	defer cancelCtx()
```

Comment:
Actually, throughout the codes, we use like below.

```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), c.cfg.NetworkTimeout)
	defer cancel()
```

Reply:
Plus, the scope of cancel context is too long. Just imagine below for-loop loops over 7 days output submission!